### PR TITLE
Support private channels

### PR DESCRIPTION
**Pull request by insomniacslk**
This patch adds support to private channels. Previously, private
channels were read-only, while now it's bidirectional like public
channels and IMs.

This required a refactoring of the channels API, and a few bug fixes.
Now anything channel-related goes through the Channel and Channels
structures, and should not use the Slack conversations API directly
anymore.

Signed-off-by: Andrea Barberio <insomniac@slackware.it>

### CODE REVIEW

**Comment by insomniacslk** [comment]
CC @gjabell

**Comment by insomniacslk** [comment]
TODO before merging this PR: now the channel names in the chanlist don't show have the prefix anymore (e.g. `general` instead of `#general`)

File: pkg/ircslack/channel.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+// IRCName returns the channel name as it would appear on IRC.
+// Examples:
+// * #channel for public groups
+// * @channel for private groups
+// * &Gxxxx|nick1-nick2-nick3 for multi-party IMs
+func (c *Channel) IRCName() string {
+	switch {
+	case c.IsPublicChannel():
+		return ChannelPrefixPublicChannel + c.Name
+	case c.IsGroup && c.IsPrivate:
```
Comment (janza) [inline pkg/ircslack/channel.go:71]: seems like  `c.IsPrivateChannel()` can be used here

Comment (insomniacslk) [reply pkg/ircslack/channel.go:71]: good catch, thanks!

Comment (insomniacslk) [reply pkg/ircslack/channel.go:71]: I'm still fixing a bunch of other things before this is ready to review, btw I've addressed your comment for the next update

**Comment by insomniacslk** [comment]
Tested on multiple teams on public and private channels, and on IM. Still not working in threads

**Comment by insomniacslk** [comment]
Now also tested on threads in public and private channels

**Comment by insomniacslk** [comment]
Messages and threads to self are not yet supported. See https://github.com/insomniacslk/irc-slack/issues/166

**Comment by insomniacslk** [comment]
Will merge this PR tomorrow if no other issue appears

**Comment by insomniacslk** [comment]
the last update fixes IMs that I broke accidentally in this PR. So now the test matrix is:

|    | public channel | private channel | multiparty IM | IM |
| --- | --- | --- | --- | --- |
| from me | works | works | untested | works |
| to me | works | works | untested | works |
| thread from me | works | works | untested | doesn't work (roadmap) |
| thread to me | untested | untested | untested | untested |

**Comment by insomniacslk** [comment]
Updated readme with feature matrix

**Comment by insomniacslk** [comment]
Merged! This one took quite a lot of work, but I'm happy with the result (hopefully I won't take those words back :D ). Next is to add some serious testing, since any change to this project is hard to validate.

**Pull request merged**